### ISSUE TITLE
Firestore emulator not supporting HTTP2 / gRPC 

### ISSUE THREAD

**Issue by nickstamas** [firebase/firebase-tools#2518]
### Environment info

**firebase-tools:** 8.6.0

**Platform:** macOS Catalina 10.15.6

### Test case

```
import firestore from '@react-native-firebase/firestore';
firestore().settings({host: 'abc123.ngrok.io', ssl: true});

...

componentDidMount = async () => {
  firestore().collection('users').add({name: 'Sam'});
}
```

Steps to reproduce

Run `firebase emulators:start`.

Run ngrok.

```
authtoken: xxxx
tunnels:
  firestore:
    addr: 8080
    proto: http
    bind-tls: true
```

Run the app with ngrok host.

### Expected behavior

A user is created in the local Firestore instance.

### Actual behavior

No user is created. After some investigation, it seems that the emulator is not using HTTP/2 and therefore not able to support the gRPC protocol required for Firestore. The emulator is running, I can modify the Firestore data through the local UI web interface, and if I hit the ngrok link from a browser I get the "Ok" response.

**Comment by samtstern** [comment]
@nickstamas we are pretty certain that the emulator supports all of the protocols needed in the Firebase SDKs, we use the emulator frequently to connect to Android, iOS, and Web apps.

It seems more likely that `ngrok` is the issue and it's not routing the Firestore traffic correctly.  Can you reproduce this issue without using `ngrok`?

**Comment by nickstamas** [comment]
@samtstern Yeah, I understand that it sounds strange. But yeah, I'm getting the exact same result with the iOS simulator and localhost. I'm wondering if it could be an issue with Catalina? Or a specific version of Catalina?

**Comment by nickstamas** [comment]
Here's an example of the output from XCode

```
2020-08-03 09:58:03.224897-0400 Reign[69109:902017] 6.27.1 - [Firebase/Firestore][I-FST000001] WriteStream (7fad0a91c178) start
2020-08-03 09:58:03.225223-0400 Reign[69109:902017] 6.27.1 - [Firebase/Firestore][I-FST000001] WriteStream (7fad0a91c178) initial request: <WriteRequest 0x700000d5e440>: {
  database: "projects/reign-dev/databases/(default)"
}
2020-08-03 09:58:03.225466-0400 Reign[69109:902017] 6.27.1 - [Firebase/Firestore][I-FST000001] GrpcStream('600003d1c120'): operation of type 2 failed
2020-08-03 09:58:03.225632-0400 Reign[69109:902017] 6.27.1 - [Firebase/Firestore][I-FST000001] GrpcStream('600003d1c120'): finishing the underlying call
2020-08-03 09:58:03.225770-0400 Reign[69109:902017] 6.27.1 - [Firebase/Firestore][I-FST000001] GrpcStream('600003d1c120'): fast finishing 1 completion(s)
2020-08-03 09:58:03.225877-0400 Reign[69109:902017] 6.27.1 - [Firebase/Firestore][I-FST000001] GrpcStream('600003d1c120'): finishing and notifying observers
2020-08-03 09:58:03.225988-0400 Reign[69109:902017] 6.27.1 - [Firebase/Firestore][I-FST000001] GrpcStream('600003d1c120'): shutting down; completions: 0, is finished: true
2020-08-03 09:58:03.226133-0400 Reign[69109:902017] 6.27.1 - [Firebase/Firestore][I-FST000001] GrpcStream('600003d1c120'): fast finishing 0 completion(s)
2020-08-03 09:58:03.923473-0400 Reign[69109:902017] 6.27.1 - [Firebase/Firestore][I-FST000001] WriteStream (7fad0a91c178) Stream error: 'Unavailable: failed to connect to all addresses'
2020-08-03 09:58:03.923668-0400 Reign[69109:902017] 6.27.1 - [Firebase/Firestore][I-FST000001] GrpcStream('600003d1c120'): destroying stream
2020-08-03 09:58:03.923790-0400 Reign[69109:902017] 6.27.1 - [Firebase/Firestore][I-FST000001] WriteStream (7fad0a91c178) backoff
2020-08-03 09:58:03.923909-0400 Reign[69109:902017] 6.27.1 - [Firebase/Firestore][I-FST000001] Backing off for 33631 ms (base delay: 60000 ms, delay with jitter: 33632 ms, last attempt: 1 ms ago)
```

**Comment by samtstern** [comment]
@nickstamas thanks for all the extra info.  Assigning to @samhorlbeck to debug.

**Comment by akauppi** [comment]
Just checking - are you aware of #2571 ?

TL;DR Firebase emulator does not do HTTP/2. Because of `superstatic` dependency.

**Comment by MeghaB** [comment]
Hey @akauppi  - #2571 is related to the Hosting emulator, not Firestore, so is likely unrelated. 

Working to see if we can reproduce this. b/169600050  for internal tracking

**Comment by akauppi** [comment]
@MeghaB Yes, noticed the same just a moment ago. Keeping my comment since also others might confuse these two.

**Comment by yuchenshi** [comment]
@nickstamas  Could you please try this with a TCP tunnel instead? I'm not familiar with ngrok, but the documentation says [TCP tunnels are usually better for databases](https://ngrok.com/docs#tcp), and the Firestore Emulator is a database.

**Comment by troyshu** [comment]
I'm having the same issue as @nickstamas did, trying to use firestore emulator with an ngrok host so that I can access firestore emulator from a test device connected to my computer. I tried @yuchenshi 's tip to use an ngrok TCP tunnel (`ngrok tcp 8080` because firestore emulator is running on port 8080), the end result is the same: no data is written to the firestore emulator (or read from it). For whatever reason, it seems an app on a device is not able to read/write from a local firestore emulator when tunned through a http or tcp tunnel like ngrok (which in my understanding is required for an app on a device to communicate with a locally run server).

Instead of interacting with the firestore emulator from a mobile device, for now I think I'll move to interacting with the firestore emulator from a mobile simulator. This will be my dev setup. Then staging will be testing the app on the mobile device itself + a staging firestore database/firebase project. Then prod will be the released mobile app + the prod firestore database/firebase project.

But I feel like the root cause here should still be identified: there are probably devs like me who will spend a lot of time trying to figure out why they can't interact with their local firestore emulator from a device, even though interacting with any other kind of local server or database works fine from a device (in my experience, also when using ngrok).

**Comment by troyshu** [comment]
The tip [in this comment](https://github.com/FirebaseExtended/flutterfire/discussions/5546#discussioncomment-557317) to turn on a VPN helped me fix this issue. My specific issue was that my Flutter app when used on a mobile device was not connecting to my _local_ Firestore emulator, emitting the same errors [here](https://github.com/FirebaseExtended/flutterfire/discussions/5546). Maybe the root cause is because I have a [multi-router config with DHCP on](https://github.com/FirebaseExtended/flutterfire/discussions/5546#discussioncomment-811301) 🤷

**Steps to my fix:**

0. Turn on VPN.

**Option 1:**

1. Tell the Firestore emulator to use host "0.0.0.0" in `firebase.json` in the directory for your Firebase project
e.g.
```
"emulators": {
    "firestore": {
      "port": 8080,
     "host": "0.0.0.0"
    },
}
``` 
2. In your Flutter app's code, use your local computer's IP address which will look like `192.168.x.xx`. e.g. `FirebaseFirestore.instance.useFirestoreEmulator('192.168.x.xx', 8080);`

**Option 2:**
1. Leave the "host" field out in `firebase.json` for your Firestore emulator
2. Use ngrok (or something similar) to tunnel _TCP_ traffic at a specific port to your local Firestore emulator address (see [tip](https://github.com/firebase/firebase-tools/issues/2518#issuecomment-722561824) here).
3. In your Flutter app's code, use the address (and port) that ngrok will give you for the TCP tunnel,  will look like `1.tcp.ngrok.io:[port]`. e.g. `FirebaseFirestore.instance.useFirestoreEmulator('1.tcp.ngrok.io', [port]);`

Remember to make sure your Firestore emulator's security rules are set up to allow the queries you're making. Also make sure that your Flutter app is connecting to the same project-id that the Firestore emulator is running for. e.g. if you're using the [Flutterfire CLI](https://firebase.flutter.dev/docs/cli/) to configure Firebase for your Flutter project, make sure the project-ids in the generated `firebase_options.dart` in your Flutter project direcotry match the project-id in the `.firebaserc` in your Firebase project's directory.

**Comment by emileswain** [comment]
Going to share my implementation as it adds in the specific detail for configuring ngrok with Firebase Emulator with Android project building to device.  I experienced all the above issues as well as some others. But the above post solved the main issue, which was I was setting ngrok to expect use of its http protocol from firebstore, but infact firestore was expecting tcp. 

I launched the emulator like so.

```
 firebase emulators:start --import=./data --export-on-exit
...
┌─────────────────────────────────────────────────────────────┐
│ ✔  All emulators ready! It is now safe to connect your app. │
│ i  View Emulator UI at http://192.168.0.29:4000             │
└─────────────────────────────────────────────────────────────┘

┌────────────────┬───────────────────┬────────────────────────────────────┐
│ Emulator       │ Host:Port         │ View in Emulator UI                │
├────────────────┼───────────────────┼────────────────────────────────────┤
│ Authentication │ 192.168.0.29:9099 │ http://192.168.0.29:4000/auth      │
├────────────────┼───────────────────┼────────────────────────────────────┤
│ Functions      │ 192.168.0.29:5001 │ http://192.168.0.29:4000/functions │
├────────────────┼───────────────────┼────────────────────────────────────┤
│ Firestore      │ 192.168.0.29:8080 │ http://192.168.0.29:4000/firestore │
└────────────────┴───────────────────┴────────────────────────────────────┘
  Emulator Hub running at localhost:4400
  Other reserved ports: 4500

```


Below is my NGrok yml file. Note that the main fix here was to set the firestore proto to tcp. 
```
authtoken: YourNgrokToaken

tunnels:
  yourProject-dev-ui:
    addr: 192.168.0.29:4000
    proto: http
    subdomain: yourProject-dev-ui
  yourProject-dev-functions:
    addr: 192.168.0.29:5001
    proto: http
    subdomain: yourProject-dev-functions
  yourProject-dev-auth:
    addr: 192.168.0.29:9099
    proto: http
    subdomain: yourProject-dev-auth
  yourProject-dev-firestore:
    addr: 192.168.0.29:8080
    proto: tcp  
    subdomain: yourProject-dev-firestore

```

Launched ngrok from same directory as my firebase cloud function project. Though obviously you can launch from anywhere.

```
& 'C:\Program Files\ngrok\ngrok.exe' start -config ngrok.config.yml --all

Forwarding                    tcp://6.tcp.ngrok.io:15252 -> 192.168.0.29:8080
Forwarding                    http://yourProject-dev-auth.ngrok.io -> http://192.168.0.29:9099
Forwarding                    https://yourProject-dev-auth.ngrok.io -> http://192.168.0.29:9099
Forwarding                    http://yourProject-dev-functions.ngrok.io -> http://192.168.0.29:5001
Forwarding                    https://yourProject-dev-functions.ngrok.io -> http://192.168.0.29:5001
Forwarding                    http://yourProject-dev-ui.ngrok.io -> http://192.168.0.29:4000
Forwarding                    https://yourProject-dev-ui.ngrok.io -> http://192.168.0.29:4000

Connections                   ttl     opn     rt1     rt5     p50     p90 
```


Then in a suitably early place for intialising FirebaseAuth, and FirebaseFireStore in my Flutter project. (i have this in main_dev.dart for my dev flavour.

```
Future<void> main() async {
...
  FirebaseAuth.instance.useAuthEmulator('yourProject-dev-auth.ngrok.io',80);//9099
  FirebaseFirestore.instance.useFirestoreEmulator('6.tcp.ngrok.io',16334);//8080
...
}
```

I had followed the above suggestions but lazily missed out the fact that i had to change the proto to tcp in my config file. 
An intermediate error i recieved, due to my half assed attempt to resolve the issue was. 
```
Status{code=INTERNAL, description=error in frame handler, cause=java.io.IOException: FRAME_SIZE_ERROR: 4740180
```
This [post](https://github.com/grpc/grpc-java/issues/8164) made me realise i was passing http as the protocol when it was actually checking for tcp. 

Anyway, i hope this helps others spend less time getting firebase up and running with Ngrok. 

Note: 
I believe that both the following worked, in terms of an android device being able to work with the emulator. As long as they are on the same network.
```
 FirebaseFirestore.instance.useFirestoreEmulator('6.tcp.ngrok.io',16334);
```
```
 FirebaseFirestore.instance.useFirestoreEmulator('192.168.0.29.',8080);
```
